react/features/recording/components/RecordingButton.web.js

```diff
@@ -0,0 +1,307 @@
+};
+
+/**
+ * React {@code Component} for displaying an inline dialog for changing receive
+ * video settings.
+ *
+ * @extends Component
+ */
+class RecordingButton extends Component {
+    /**
+     * {@code RecordingButton}'s property types.
+     *
+     * @static
+     */
+    static propTypes = {
+        /**
+         * The redux store representation of the JitsiConference.
+         */
+        _conference: React.PropTypes.object,
+
+        /**
+         * Indicates if the recording button has been toggled.
+         */
+        _recordingButtonToggled: React.PropTypes.bool,
+
+        /**
+         * Indicates if the audio recording option should be enabled.
+         */
+        _recordingEnabled: React.PropTypes.bool,
+
+        /**
+         * Indicates if the transcription option should be enabled.
+         */
+        _transcriptionEnabled: React.PropTypes.bool,
+
+        /**
+         * Whether or not the button is visible, based on the visibility of the
+         * toolbar. Used to automatically hide the inline dialog if not visible.
+         */
+        _visible: React.PropTypes.bool,
+
+        /**
+         * Invoked to obtain translated string.
+         */
+        t: React.PropTypes.func,
+
+        /**
+         * From which side tooltips should display. Will be re-used for
+         * displaying the inline dialog for video quality adjustment.
+         */
+        tooltipPosition: React.PropTypes.string
+    };
+
+    /**
+     * Initializes a new {@code RecordingButton} instance.
+     *
+     * @param {Object} props - The read-only properties with which the new
+     * instance is to be initialized.
+     */
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            /**
+             * Whether or not the inline dialog for adjusting received video
+             * quality is displayed.
+             */
+            showRecordingDialog: false,
+
+            isToggleRecordingChecked: false,
+
+            isToggleTranscriptionChecked: false
+        };
+
+        // Bind event handlers so they are only bound once for every instance.
+        this._onDialogClose = this._onDialogClose.bind(this);
+        this._onDialogToggle = this._onDialogToggle.bind(this);
+        this._onToggleRecordingChange
+            = this._onToggleRecordingChange.bind(this);
+        this._onToggleTranscriptionChange
+            = this._onToggleTranscriptionChange.bind(this);
+    }
+
+    /**
+     * Updates the toggled state depending on the _recordingButtonToggled
+     * prop.
+     *
+     * @param {Object} nextProps - The props that will be applied after the
+     * update.
+     * @inheritdoc
+     * @returns {void}
+     */
+    componentWillUpdate(nextProps) {
+        if (nextProps._recordingButtonToggled
+            !== this.props._recordingButtonToggled
+            && nextProps._recordingButtonToggled
+            !== this.state.isToggleRecordingChecked) {
+            this.setState({
```

Comment:
Does this this.setState work without issue? The react docs say this:

> Note that you cannot call this.setState() here. If you need to update state in response to a prop change, use componentWillReceiveProps() instead.

Reply:
Oops, well, I'll need to fix this then. I haven't noticed any issues with it, but haven't tested that much either. Also if the doc is so clear about it I'll definitely change it.